gui/journal.lua

```diff
@@ -0,0 +1,703 @@
+            self.cursor.y
+        )
+        return true
+    elseif keys.CUSTOM_CTRL_A then
+        -- line start
+        self:setCursor(1, self.cursor.y)
+        return true
+    elseif keys.CUSTOM_CTRL_E then
+        -- line end
+        self:setCursor(
+            #self.lines[self.cursor.y],
+            self.cursor.y
+        )
+        return true
+    elseif keys.CUSTOM_CTRL_U then
+        -- delete current line
+        if (self:hasSelection()) then
+            -- delete all lines that has selection
+            self:setSelection(
+                1,
+                self.cursor.y,
+                #self.lines[self.sel_end.y],
+                self.sel_end.y
+            )
+            self:eraseSelection()
+        else
+            local y = self.cursor.y
+            self:setSelection(1, y, #self.lines[y], y)
+            self:eraseSelection()
+
+            -- local line_start = self:cursorToIndex(1, self.cursor.y)
+            -- local line_end = self:cursorToIndex(#self.lines[self.cursor.y], self.cursor.y)
+            -- local new_text = self.text:sub(1, line_start - 1) .. self.text:sub(line_end + 1)
+            -- self:setText(new_text)
+        end
+        return true
+    elseif keys.CUSTOM_CTRL_D then
+        -- delete char, there is no support for `Delete` key
+        local old = self.text
+        if (self:hasSelection()) then
+            self:eraseSelection()
+        else
+            local del_pos = self:cursorToIndex(
+                self.cursor.x,
+                self.cursor.y
+            )
+            self:setText(old:sub(1, del_pos-1) .. old:sub(del_pos+1))
+        end
+
+        return true
+    elseif keys.CUSTOM_CTRL_W then
+        -- delete one word backward
+        local ind = self:cursorToIndex(self.cursor.x, self.cursor.y)
+        local _, prev_word_end = self.text
+            :sub(1, ind-1)
+            :find('.*%s[^%s]')
+        local word_start = prev_word_end or 1
+        local cursor = self:indexToCursor(word_start - 1)
+        local new_text = self.text:sub(1, word_start - 1) .. self.text:sub(ind)
+        self:setText(new_text, cursor.x, cursor.y)
+        return true
+    elseif keys.CUSTOM_CTRL_C then
+        self:copy()
+        return true
+    elseif keys.CUSTOM_CTRL_X then
+        self:cut()
+        return true
+    elseif keys.CUSTOM_CTRL_V then
+        self:paste()
+        return true
+    end
+
+end
+
+JOURNAL_PERSIST_KEY = 'dfjournal-content'
+
+JournalScreen = defclass(JournalScreen, gui.ZScreen)
+JournalScreen.ATTRS {
+    frame_title = 'journal',
+    frame_width = 80,
+    frame_height = 40,
+    focus_path='journal',
+}
+
+function JournalScreen:init()
+    local content = dfhack.persistent.getSiteDataString(JOURNAL_PERSIST_KEY) or ''
+
+    local function on_text_change(text)
+        if dfhack.isWorldLoaded() then
+            dfhack.persistent.saveSiteDataString(JOURNAL_PERSIST_KEY, text)
+        end
```

Comment:
I think we need a little more functionality around data loss prevention. `dfhack.persisitent` doesn't actually persist to disk until the game is saved. If the game crashes before then, then edits will be lost.

I'd suggest also writing to a swap file in `dfhack-config/journal.json` and timestamping the persisted data so you know if the swap file is more recent than the data loaded from `getSiteDataString`. Then you can offer to restore lost text when the game crashes and players restart `gui/journal`.

knowing when the game has been successfully saved so you can clean up swap files might be tricky. we might need to create a state change event for that that the script can hook.

Reply:
I am not sure what would be expected behaviour for the users. 

When I do not save the game, I lost the game state and I expecting it. 
The game have "Quit without saving" menu option - and I expecting that my journal would not be saved if I use it.

But I am power user, so my expectation can be differ than most of players.
---

Maybe we can ask people about it on discord or smth.
Whatever decission would be here - maybe it can be improvement for next versions?

Reply:
more thoughts: this can get tricky if a game is loaded, but the swap data is from a different savegame. (or a previous save in the same world).

alternate idea: the swap file can only hold a single entry, but will warn if the swap data doesn't match the loaded data. players will then be given the choice to use the data loaded from the savegame or the data loaded from the swap file, and whether to clear the swap file. if the swap data is unwanted for the current game, the player is encouraged to either clear the swap data or load the game where it is relevant, recover the swap there, save the game, and clear the swap.

that sounds complicated too, though, so let's keep this whole data loss prevention thing for a future PR. it might need more discussion too.